CIP-0141? | Web-Wallet Bridge - Plutus wallets (#798)
cardano-foundation/CIPs: CIP-0141/README.md

}
```

### Aditional Error Types

#### CompletedTxError

```ts
CompletedTxErrorCode = {
	NotFound: 1,
	NotReady: 2
}
```

* NotFound - The transaction with the given id was not found.
* NotReady - The transaction with the given id is not ready yet. 

### V2 Additional API Endpoints

#### `api.cipxxxx.getScriptRequirements`: Promise<ScriptRequirement[]>

Errors: `APIError`

Returns a list of `ScriptRequirement` that will be used to validate any transaction sent to the wallet. Every field in the `ScriptRequirement` object is optional, and the wallet should only include the fields that are relevant to the current transaction.

For wallets with multiple spend conditions, separate entries in the list should be used to represent each spend condition. Wallet providers should implement UX to allow users to order the list of `ScriptRequirement` from most to least preferred. DApps should use the first entry in the list that is valid for the current transaction or select one based on the logic of their use-case.

#### `api.cipxxxx.getScript()`: Promise[<number<plutusVersion>>,<CBOR<plutusScript>>]

Errors: `APIError`

Returns the CBOR-encoded Plutus script that controls this wallet.

> #### `api.cipxxxx.submitUnsignedTx(tx: CBOR<unsignedTransaction>)`: Promise<hash32>

Reviewer: ```suggestion
#### `api.cip141.submitUnsignedTx(tx: CBOR<unsignedTransaction>)`: Promise<hash32>
```

If the transaction contains hidden metadata, the wallet should not submit the transaction when it is ready, but return it to the DApp when the DApp calls the `getCompletedTx` function.

It is expected that this will be the endpoint used by all wallets that require multiple signatures to sign a transaction.

#### `api.cipxxxx.getCompletedTx(txId: hash32)`: Promise[<CBOR<transaction>,CBOR<transaction_witness_set>>]

Errors: `APIError`, `CompletedTxError`

If the transaction is not ready, the wallet should throw a `CompletedTxError` with the appropriate error code. If the transaction is ready, the wallet should return the CBOR-encoded transaction and the signatures.

> #### `api.cipxxxx.getSecret(secretId: String)`: Promise<String>

Reviewer: ```suggestion
#### `api.cip141.getSecret(secretId: String)`: Promise<String>
```

Errors: `APIError`

Returns the secret associated with the given secretId. The secret should be returned as a hex-encoded string. 

Wallets should request authorization from the user before returning the secret to the DApp.

> #### `api.cipxxxx.signRedeemer(data: Data, primitive : encriptionPrimitive  )`: Promise<CBOR<redeemer>>

Reviewer: ```suggestion
#### `api.cip141.signRedeemer(data: Data, primitive : encriptionPrimitive  )`: Promise<CBOR<redeemer>>
```

Errors: `APIError`
Returns the CBOR-encoded redeemer.

### Altered API endpoints

#### `api.signTx(tx: CBOR<transaction>, partialSign: bool = false)`: Promise<CBOR<transaction_witness_set>>

This endpoint will now optionally return an error if the smart Wallet is a multiparty schema and signatures need to be gathered from multiple parties asynchronously.

#### `api.signData(addr: Address, payload: Bytes)`: Promise<DataSignature>

Plutus contracts cannot sign data, only validate transactions. This endpoint will be disabled when connecting to a wallet using this extension.

## Rationale: how does this CIP achieve its goals?

By altering the API endpoints and adding new ones, we can provide the necessary information for DApps to interact with Plutus wallets. This will allow any developer to integrate smart-wallets into their DApps, while providing a consistent interface for all wallets.

## Path to Active

### Acceptance Criteria

> - [ ] The interface is implemented and supported by various wallet providers.

Reviewer: ```suggestion
- [ ] The interface is implemented and supported by three different wallet providers.
```
